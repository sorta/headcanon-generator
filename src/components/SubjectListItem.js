import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { allowEdit } from '../utils'

class SubjectListItem extends Component {
  static propTypes = {
    subject: PropTypes.shape({
      name: PropTypes.string,
      fandomKey: PropTypes.string,
    }),
    updateSubject: PropTypes.func,
    deleteSubject: PropTypes.func,
    index: PropTypes.string,
    setAvailability: PropTypes.func,
    subjectAvailability: PropTypes.object,
    fandomAvailability: PropTypes.object,
    isManaging: PropTypes.func,
  };
  static defaultProps = {
    subject: { name: '', fandomKey: '' },
    updateSubject: () => {},
    deleteSubject: () => {},
    index: '',
    setAvailability: () => {},
    subjectAvailability: {},
    fandomAvailability: {},
    isManaging: () => false,
  };

  handleChange = (event) => {
    const updatedSubject = {
      ...this.props.subject,
      [event.currentTarget.name]: event.currentTarget.value,
    };

    this.props.updateSubject(this.props.index, updatedSubject);
  };

  toggleAvailability = (event) => {
    this.props.setAvailability('subject', this.props.index, event.currentTarget.checked);
  };

  render() {
    let availCheckbox = null;
    if (!({}.hasOwnProperty.call(this.props.fandomAvailability, this.props.subject.fandomKey))) {
      availCheckbox = (
        <input
          id={`isAvailable_subj_${this.props.index}`}
          name="isAvailable"
          type="checkbox"
          className="cb-fake"
          checked={!({}.hasOwnProperty.call(this.props.subjectAvailability, this.props.index))}
          onChange={this.toggleAvailability}
        />
      );
    }

    if (!this.props.isManaging() || !allowEdit) {
      return (
        <li key={this.props.index} className="SubjectListItem list-item control-row">
          {availCheckbox}
          <label
            htmlFor={`isAvailable_subj_${this.props.index}`}
            className={availCheckbox === null ? "SubjectListItem-text strike" : "SubjectListItem-text"}
          >
            {this.props.subject.name}
          </label>
        </li>
      );
    }

    return (
      <li key={this.props.index} className="SubjectListItem list-item control-row">
        {availCheckbox}
        <label htmlFor={`isAvailable_subj_${this.props.index}`}></label>
        <input name="name" type="text" placeholder="Subject Name" className="control-stretch"
          value={this.props.subject.name} onChange={this.handleChange}
        />
        <button
          className="SubjectListItem-delete"
          onClick={() => this.props.deleteSubject(this.props.index)}
        >-</button>
      </li>
    );
  }
}

export default SubjectListItem;
