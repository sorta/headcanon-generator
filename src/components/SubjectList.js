import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SubjectListItem from './SubjectListItem';

class SubjectList extends Component {
  static propTypes = {
    subjects: PropTypes.object,
    subjectKeys: PropTypes.array,
    updateSubject: PropTypes.func,
    deleteSubject: PropTypes.func,
    isManaging: PropTypes.func,
    setAvailability: PropTypes.func,
    subjectAvailability: PropTypes.object,
    fandomAvailability: PropTypes.object,
  };
  static defaultProps = {
    subjects: {},
    subjectKeys: [],
    updateSubject: () => {},
    deleteSubject: () => {},
    isManaging: () => false,
    setAvailability: () => {},
    subjectAvailability: {},
    fandomAvailability: {},
  };

  render() {
    if (Object.keys(this.props.subjects).length === 0) {
      return null;
    }

    return (
      <ul className="options-list SubjectList">
        {this.props.subjectKeys.sort().map(key => {
          if ({}.hasOwnProperty.call(this.props.subjects, key)) {
            return (
              <SubjectListItem
                key={key}
                index={key}
                subject={this.props.subjects[key]}
                updateSubject={this.props.updateSubject}
                deleteSubject={this.props.deleteSubject}
                isManaging={this.props.isManaging}
                setAvailability={this.props.setAvailability}
                subjectAvailability={this.props.subjectAvailability}
                fandomAvailability={this.props.fandomAvailability}
              />
            );
          } else {
            return null;
          }
        })}
      </ul>
    );
  }
}

export default SubjectList;
