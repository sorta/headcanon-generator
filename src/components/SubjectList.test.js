import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import SubjectList from './SubjectList';
import SubjectListItem from './SubjectListItem';

const props = {
  subjects: {
    s_test_1: { name: 'Test 1' },
    s_test_2: { name: 'Test 2' },
    s_test_3: { name: 'Test 3' },
  },
  subjectAvailability: {},
  updateSubject: jest.fn(),
  deleteSubject: jest.fn(),
  setAvailability: jest.fn(),
  isManaging: () => false,
  subjectKeys: [
    's_test_1',
    's_test_2',
    's_test_3',
  ],
  fandomAvailability: {},
};

// Tests for regular mode
describe("SubjectList", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <SubjectList { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a SubjectList ul element', () => {
    const component = shallow(<SubjectList { ...props } />);
    expect(component.find('ul.SubjectList.options-list').exists()).toBe(true);
  });

  it('has right number of SubjectListItems', () => {
    const component = shallow(<SubjectList { ...props } />);
    expect(component.find(SubjectListItem)).toHaveLength(props.subjectKeys.length);
  });
});
