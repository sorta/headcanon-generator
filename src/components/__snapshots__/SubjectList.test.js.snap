// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectList renders and matches our snapshot 1`] = `
<ul
  className="options-list SubjectList"
>
  <li
    className="SubjectListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_subj_s_test_1"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="SubjectListItem-text"
      htmlFor="isAvailable_subj_s_test_1"
    >
      Test 1
    </label>
  </li>
  <li
    className="SubjectListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_subj_s_test_2"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="SubjectListItem-text"
      htmlFor="isAvailable_subj_s_test_2"
    >
      Test 2
    </label>
  </li>
  <li
    className="SubjectListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_subj_s_test_3"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="SubjectListItem-text"
      htmlFor="isAvailable_subj_s_test_3"
    >
      Test 3
    </label>
  </li>
</ul>
`;
